### PR TITLE
Respect DOH Dns config in IPFS

### PR DESCRIPTION
**Pull request by cypt4**
Resolves https://github.com/brave/brave-browser/issues/22409
DOH url is taken from network service request. After DNS config updates IPFS daemon restarts.
Sec review: https://github.com/brave/security/issues/949

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
1) Check ipfs opens pages.
2) Enable DOH.
3) Check ipfs opens pages and check that there are no default dns requests to ipfs nodes.
4) Write custom DNS server and check that ipfs still works.

### CODE REVIEW

File: browser/ipfs/ipfs_dns_resolver_impl.cc [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+void IpfsDnsResolverImpl::OnDnsConfigChangeManagerConnectionError() {
+  NOTREACHED() << "DNS Config change manager connection error.";
+}
+
+void IpfsDnsResolverImpl::OnDnsConfigChanged() {
+  Notify(GetDnsOverHttps());
+}
+
+absl::optional<std::string> IpfsDnsResolverImpl::GetDnsOverHttps() {
```
Comment (yrliou) [inline browser/ipfs/ipfs_dns_resolver_impl.cc:43]: nit: `GetDnsOverHttpsServers`

File: browser/ipfs/ipfs_dns_resolver_impl.cc [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  SecureDnsConfig secure_dns_config =
+      SystemNetworkContextManager::GetStubResolverConfigReader()
+          ->GetSecureDnsConfiguration(
+              false /* force_check_parental_controls_for_automatic_mode */);
+  const auto servers = secure_dns_config.doh_servers().servers();
+  if (secure_dns_config.mode() == net::SecureDnsMode::kOff || servers.empty()) {
+    return absl::nullopt;
+  }
+  std::string server_template = servers[0].server_template();
```
Comment (yrliou) [inline browser/ipfs/ipfs_dns_resolver_impl.cc:52]: Why do we only pass the first server to ipfs config? Doesn't DNS.Resolvers config support multiple?

File: components/ipfs/ipfs_service.h [hunk 3]
```diff
@@ -216,6 +222,7 @@ class IpfsService : public KeyedService,
   base::ObserverList<IpfsServiceObserver> observers_;
 
   PrefService* prefs_ = nullptr;
+  PrefChangeRegistrar prefs_change_registrar_;
```
Comment (yrliou) [inline components/ipfs/ipfs_service.h:225]: I'm not seeing it being used?

File: browser/ipfs/ipfs_dns_resolver_impl.cc [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      receiver_.BindNewPipeAndPassRemote());
+  receiver_.set_disconnect_handler(base::BindOnce(
+      &IpfsDnsResolverImpl::OnDnsConfigChangeManagerConnectionError,
+      base::Unretained(this)));
+}
+
+IpfsDnsResolverImpl::~IpfsDnsResolverImpl() {}
+
+void IpfsDnsResolverImpl::OnDnsConfigChangeManagerConnectionError() {
+  NOTREACHED() << "DNS Config change manager connection error.";
```
[the remark this answers is not in the document]
Comment (cypt4) [reply browser/ipfs/ipfs_dns_resolver_impl.cc]: Added reconnect try here

File: browser/ipfs/ipfs_dns_resolver_impl.cc (hunk 2 above)
Comment (cypt4) [reply browser/ipfs/ipfs_dns_resolver_impl.cc:63]: https://github.com/ipfs/kubo/issues/8173#issuecomment-1100210145

Comment (yrliou) [reply browser/ipfs/ipfs_dns_resolver_impl.cc:63]: Thanks I understand it now, please rename as GetDnsOverHttpsServer or GetFirstDnsOverHttpsServer as we will always return only the first doh server.

**Review by yrliou — APPROVED** [review]
++ if https://github.com/brave/brave-core/pull/14068#discussion_r920753124 is addressed and CI is green.

**Pull request merged**